lib/facts/macosx/virtual.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+module Facts
+  module Macosx
+    class Virtual
+      FACT_NAME = 'virtual'
+
+      def call_the_resolver
+        fact_value = check_vmware || check_virtualbox || check_parallels
+
+        Facter::ResolvedFact.new(FACT_NAME, fact_value)
+      end
+
+      private
+
+      def check_vmware
+        model_identifier = Facter::Resolvers::Macosx::SystemProfiler.resolve(:model_identifier)
+        return 'vmware' if model_identifier&.start_with?('VMware')
+      end
+
+      def check_virtualbox
+        boot_rom_version = Facter::Resolvers::Macosx::SystemProfiler.resolve(:boot_rom_version)
+        return 'virtualbox' if boot_rom_version&.start_with?('VirtualBox')
+      end
+
+      def check_parallels
+        subsystem_vendor_id = Facter::Resolvers::Macosx::SystemProfiler.resolve(:subsystem_vendor_id)
+        return 'parallels' if subsystem_vendor_id&.start_with?('0x1ab8')
```

Comment:
Is **0x1ab8** constant? Won't it change on other versions?

Reply:
This is a constant used since Facter 2.x https://github.com/puppetlabs/facter/blob/f81a7066c98e2c47e0f9f9f17b709ccf34cbf52e/lib/facter/virtual.rb#L59

It is the same for Facter 3.x https://github.com/puppetlabs/facter/blob/56309e95024a5617c95296a309a80770e2ae178e/lib/src/facts/osx/virtualization_resolver.cc#L37